### PR TITLE
feat(toFilled): add toFilled

### PR DESCRIPTION
**Pull request by wondonghwi**
close: #118 

The fill function was added to es-toolkit, and since fill mutates the original array, a new function that does not modify the original array was needed. To address this, we added a new function, toFilled.

I have added a function called toFilled that does not modify the original array. This function fills a specified portion of an array with a given value but returns a new array instead of modifying the original one. Additionally, this function allows for the use of negative indices, meaning you can specify positions from the end of the array backward. The toFilled function works similarly to the fill function but stands out by not altering the original array.

In the benchmark results, since lodash does not have a toFilled function, I compared it with lodash’s fill function. The comparison showed that the toFilled function performed better in terms of performance.

The implementation of this function was inspired by [this issue](https://github.com/toss/es-toolkit/issues/118).
스크린샷 2024-07-10 오후 11 36 13

### CODE REVIEW

File: benchmarks/toFilled.bench.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { bench, describe } from 'vitest';
+import { fill as fillLodash } from 'lodash';
+import { fill as toFilled } from 'es-toolkit';
```
Comment (jgjgill) [inline benchmarks/toFilled.bench.ts:3]: I think `toFilled as toFilledToolkit`, not `fill as toFilled`.

```suggestion
import { toFilled as toFilledToolkit } from 'es-toolkit';
```

Comment (end-jarry) [reply benchmarks/toFilled.bench.ts]: Thank you. I have made the corrections. 🙃

Comment (wondonghwi) [reply benchmarks/toFilled.bench.ts]: Thank you. I have made the corrections. 🙃

File: docs/ko/reference/array/toFilled.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# toFilled
+
+배열의 특정 요소를 지정된 값으로 채운 새로운 배열을 생성해요.
+이 함수는 원본 배열을 변경하지 않아요.
+
+주어진 배열을 기반으로 새로운 배열을 생성하고, 시작 위치부터 끝 위치까지의 요소를 지정된 값으로 채워요. 시작 위치 또는 끝 위치를 제공하지 않으면 배열 전체를 채워요.
+
+음수 인덱스를 사용할 수도 있으며, 이 경우 배열의 끝에서부터 인덱스를 계산해요.
```
Comment (raon0211) [inline docs/ko/reference/array/toFilled.md:8]: 배열의 요소를 지정된 값으로 채워요. 시작 위치부터 끝 위치까지의 요소들을 제공된 값으로 대체한 새로운 배열을 반환해요. 시작 또는 끝 인덱스를 제공하지 않으면 배열 전체를 채워요.

Comment (raon0211) [inline docs/ko/reference/array/toFilled.md:8]:
```suggestion
배열의 요소를 지정된 값으로 채워요. 시작 위치부터 끝 위치까지의 요소들을 제공된 값으로 대체한 새로운 배열을 반환해요. 시작 또는 끝 인덱스를 제공하지 않으면 배열 전체를 채워요.
```

File: docs/ko/reference/array/toFilled.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+## 인터페이스
+
+```typescript
+export function toFilled<T, U>(arr: T[], value: U): Array<T | U>;
+export function toFilled<T, U>(arr: T[], value: U, start: number): Array<T | U>;
+export function toFilled<T, U>(arr: T[], value: U, start: number, end: number): Array<T | U>;
```
Comment (raon0211) [inline docs/ko/reference/array/toFilled.md:15]:
```suggestion
function toFilled<T, U>(arr: T[], value: U): Array<T | U>;
function toFilled<T, U>(arr: T[], value: U, start: number): Array<T | U>;
function toFilled<T, U>(arr: T[], value: U, start: number, end: number): Array<T | U>;
```

File: docs/ko/reference/array/toFilled.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+```
+
+### 파라미터
+
+- `arr` (`Array<T>`): 새 배열의 기반이 되는 배열입니다.
+- `value` (`U`): 새 배열을 채울 값입니다.
+- `start` (`number, 기본값 = 0`): 시작 위치입니다. 기본값은 0입니다.
+- `end` (`number, 기본값 = array.length`): 끝 위치입니다. 기본값은 배열의 길이입니다.
```
Comment (raon0211) [inline docs/ko/reference/array/toFilled.md:23]:
```suggestion
- `arr` (`Array<T>`): 채울 배얼이에요.
- `value` (`U`): 배열을 채울 값이에요.
- `start` (`number, 기본값 = 0`): 시작 위치예요. 기본값은 0이에요.
- `end` (`number, 기본값 = array.length`): 끝 위치예요. 기본값은 배열의 길이예요.
```

File: docs/ko/reference/array/toFilled.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+### 반환 값
+
+(`Array<T | U>`): 지정된 값으로 채워진 새로운 배열입니다.
```
Comment (raon0211) [inline docs/ko/reference/array/toFilled.md:27]:
```suggestion
(`Array<T | U>`): 지정된 값으로 채워진 새로운 배열이에요.
```

File: docs/reference/array/toFilled.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# toFilled
+
+Creates a new array filled with the specified value at specified positions. This function does not mutate the original array.
+
+This function generates a new array based on the given array and fills the elements from the start position to the end position with the specified value. If the start or end positions are not provided, the entire array is filled.
+
+Negative indices can also be used, in which case they are counted from the end of the array.
```
Comment (raon0211) [inline docs/reference/array/toFilled.md:7]:
```suggestion
Creates a new array filled with a specified value from the start position up to, but not including, the end position.

If the start or end indices are not provided, it defaults to filling the entire array.
```

File: docs/reference/array/toFilled.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+## Interface
+
+```typescript
+export function toFilled<T, U>(arr: T[], value: U): Array<T | U>;
+export function toFilled<T, U>(arr: T[], value: U, start: number): Array<T | U>;
+export function toFilled<T, U>(arr: T[], value: U, start: number, end: number): Array<T | U>;
```
Comment (raon0211) [inline docs/reference/array/toFilled.md:14]:
```suggestion
function toFilled<T, U>(arr: T[], value: U): Array<T | U>;
function toFilled<T, U>(arr: T[], value: U, start: number): Array<T | U>;
function toFilled<T, U>(arr: T[], value: U, start: number, end: number): Array<T | U>;
```

**Comment by raon0211** [comment]
Thanks for your great contribution!

I wanted to discuss the behavior of `toFilled` in comparison to the original `fill` function.

One key difference I've noticed is that `fill` supports negative indices, whereas `toFilled` does not. Since lodash's `fill` function does not support negative indices, I believe `toFilled` should follow and not support them either.

What are your thoughts on this?

**Comment by wondonghwi** [comment]
Hello @raon0211,

Based on my observations and tests, it appears that Lodash's fill function does indeed support negative indices. Given this, it seems appropriate that toFilled, as well as esToolkit's fill function, should also support negative indices to maintain consistency. Here is an example that demonstrates the behavior:

image

Given this behavior, I believe it would be beneficial to enhance both toFilled and esToolkit's fill functions to support negative indices. I would be interested in working on this enhancement if it is deemed necessary. What are your thoughts on this?

File: docs/ko/reference/array/toFilled.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
+# toFilled
+
+배열의 요소를 지정된 값으로 채워요. 시작 위치부터 끝 위치까지의 요소들을 제공된 값으로 대체한 새로운 배열을 반환해요. 시작 또는 끝 인덱스를 제공하지 않으면 배열 전체를 채워요.
```
Comment (raon0211) [inline docs/ko/reference/array/toFilled.md:3]:
```suggestion
배열의 요소를 지정된 값으로 채워요. 시작 위치부터 끝 위치까지의 요소들을 제공된 값으로 대체한 새로운 배열을 반환해요. 시작 또는 끝 인덱스를 제공하지 않으면 배열 전체를 채워요.

음수 인덱스를 사용할 수도 있어요. 이 경우 배열의 끝에서부터 인덱스를 계산해요.
```

File: docs/reference/array/toFilled.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
+# toFilled
+
+Creates a new array filled with a specified value from the start position up to, but not including, the end position.
+
+If the start or end indices are not provided, it defaults to filling the entire array.
```
Comment (raon0211) [inline docs/reference/array/toFilled.md:5]:
```suggestion
If the start or end indices are not provided, it defaults to filling the entire array.

Negative indices can also be used, in which case they are counted from the end of the array.
```

**Review by raon0211 — APPROVED** [review]
Looks good to me!

**Comment by raon0211** [comment]
>Given this behavior, I believe it would be beneficial to enhance both toFilled and esToolkit's fill functions to support negative indices.

Got this. Could you fix the behavior for `fill`, too?

**Comment by end-jarry** [comment]
@raon0211 

Sure :)

 I will raise this issue and proceed to modify the fill function accordingly!

**Comment by wondonghwi** [comment]
@raon0211 

Sure :)

 I will raise this issue and proceed to modify the fill function accordingly!

**Pull request merged**